lib/medtronic600/NGPHistoryParser.js

```diff
@@ -0,0 +1,894 @@
+              interstitialSignal: reading.isig,
+            });
+          this.addBtUtcFields(record, reading.timestamp);
+          events.push(record.done());
+        }
+      }
+    }
+
+    return this;
+  }
+
+  buildBGRecords(events) {
+    for (const event of this.eventsOfType(NGPHistoryEvent.EVENT_TYPE.BG_READING)) {
+      let reading = null;
+      if (event.bgSource === NGPUtil.NGPConstants.BG_SOURCE.SENSOR_CAL ||
+        event.calibrationFlag) {
+        reading = this.cfg.builder.makeDeviceEventCalibration()
+          .with_units('mg/dL')
+          .with_subType('calibration')
+          .with_value(event.bgValue);
+      } else {
+        reading = this.cfg.builder.makeSMBG()
+          .with_units('mg/dL')
+          .with_subType(event.bgLinked ? 'linked' : 'manual')
+          .with_value(event.bgValue);
+      }
+
+      if (event.bgLinked) {
+        reading.with_payload({
+          meterSerial: event.meterSerialNumber,
+        });
+      }
+
+      this.addBtUtcFields(reading, event.timestamp);
+
+      events.push(reading.done());
+    }
+
+    return this;
+  }
+
+  buildBasalRecords(events) {
+    for (const event of this.eventsOfType(NGPHistoryEvent.EVENT_TYPE.BASAL_SEGMENT_START)) {
+      const basal = this.cfg.builder.makeScheduledBasal()
+        .with_rate(event.rate)
+        .with_scheduleName(event.patternName);
+
+      this.addBtUtcFields(basal, event.timestamp);
+
+      events.push(basal);
+    }
+
+    return this;
+  }
+
+  buildTempBasalRecords(events) {
+    for (const event of this.eventsOfType(NGPHistoryEvent.EVENT_TYPE.TEMP_BASAL_COMPLETE)) {
+      // Since we're using the TEMP_BASAL_COMPLETE event, we need to get the Device timestamp
+      // for the *beginning* of the event.
+      const beginTimestamp = new NGPUtil.NGPTimestamp(event.timestamp.rtc - event.duration,
+        event.timestamp.offset);
+
+      const suppressedBasal = this.findSuppressedBasal(event);
+
+      // TODO - add adjustTempBasalDuration() method for when canceled() is true!
+      const basal = this.cfg.builder.makeTempBasal()
+        .with_rate(event.rate)
+        .with_duration(event.duration * sundial.SEC_TO_MSEC)
+        .set('suppressed', {
+          type: 'basal',
+          deliveryType: 'scheduled',
+          rate: suppressedBasal.rate,
+          scheduleName: suppressedBasal.patternName,
+        });
+
+      // We need to find the normal rate here to calculate event.rate, because
+      // if it's a PERCENT record, event.rate will be 0.
+      if (event.type === NGPUtil.NGPConstants.TEMP_BASAL_TYPE.PERCENT) {
+        basal
+          .with_percent(event.percentageOfRate / 100.0)
+          .with_rate(suppressedBasal.rate * (event.percentageOfRate / 100.0));
+      }
+
+      this.addBtUtcFields(basal, beginTimestamp);
+
+      events.push(basal);
+
+      // Now that the temp basal is complete, tack on the original basal schedule.
+      const followUpBasal = this.cfg.builder.makeScheduledBasal()
```

Comment:
I thought that there wasn't, because I was getting some strange results, but it looks like you are correct. The strange results were because I don't have code yet for cancelled temp basals. I'll go fix that now and remove the `followUpBasal`s. Thanks!